File: diagnoal_matrix.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
#         1==>up
#         0==>down
#         if matrix is null
        if not mat: 
            return []
        direction=1
        res=[]    
#         length of the matrix
        n=len(mat)
        m=len(mat[0]) 
        i=0
        row=0
        column=0      
        while(i<=(m*n)-1):
            res.append(mat[row][column])
            i=i+1
            if direction==1:
# this if condition explicitly checks if the column reaches out of bounds when row==0
#             if column length equals to mat[0]
                if column==m-1:
                    row=row+1
                    direction=0
                    #                 check if row==0
                elif row==0:
                    column=column+1
                    direction=0
                else:
                    column=column+1
                    row=row-1

#         if direction is 0   
            else:
        #                if row length is len(mat) i.e n
                if row==n-1:
                    column=column+1
                    direction=1
                elif column==0:
                    row=row+1
                    direction=1
                else:
                    row=row+1
                    column=column-1
        return res         
                

        
        
        
        

File: DiagnoalMatrix.java
// Time Complexity :0(mn)
// Space Complexity : 0(1)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this : no

/*
 * create a flag for the directions up and down
 * for top the general condition is row -- and column ++
 * handle the boundries
 * similary for downward direction with row++ and column-- also 
 * handle the boundries
 * */

class Solution {
	public int[] findDiagonalOrder(int[][] mat) {

		if (mat == null || mat.length == 0)
			return new int[] {};

		int m = mat.length;
		int n = mat[0].length;
		// initate m*n res arrat
		int[] res = new int[m * n];
		int r = 0;
		int c = 0;
		int dir = 1;
		int i = 0;

		while (i < res.length) {
			res[i] = mat[r][c];
			i++;
			// moving up
			if (dir == 1) {
				// boundry condition moving up and in the last column then increment the row and
				// change the directin
				if (c == n - 1) {
					r++;
					dir = -1;

				} else if (r == 0) { // boundry condition moving up and in the first row then increment the column
										// and change the directin
					c++;
					dir = -1;
				} else {
					// general case for moving up row is decresing and column is increaing
					r--;
					c++;
				}
			} else {
				// boundry condition moving down and in the last row then increment the column
				// and change the directin
				if (r == m - 1) {
					c++;
					dir = 1;
				} else if (c == 0) {
					// boundry condition moving down and in the first column then increment the row
					// and change the directin
					r++;
					dir = 1;
				} else {
					// general case for moving dow column is decresing and row is increaing
					c--;
					r++;
				}
			}

		}

		return res;
	}
}